imap_processing/ultra/l2/ultra_l2.py

```diff
@@ -0,0 +1,483 @@
+    ----------
+    l1c_products : list[xr.Dataset]
+        List of xarray Datasets containing the l1c products to combine.
+    spacing_deg : float, optional
+        The spacing of the grid in degrees, by default DEFAULT_SPACING_DEG.
+
+    Returns
+    -------
+    tuple[NDArray]
+        Tuple of the following arrays:
+        - Exposure time (combined across l1c products) for 45 degree sensor
+        - Exposure time (combined across l1c products) for 90 degree sensor
+        - Exposure time (combined across l1c products) for both sensors added together
+    """
+    num_dps_pointings = len(l1c_products)
+    (num_el_bins_l1c, num_az_bins_l1c, num_energy_bins_l1c) = l1c_products[
+        0
+    ].counts.shape
+    logger.info(
+        f"Number of DPS Pointings: {num_dps_pointings}. "
+        f"Shape is {num_el_bins_l1c} el x {num_az_bins_l1c} "
+        f"az x {num_energy_bins_l1c} energy"
+    )
+
+    # Setup the combined-across-pointings exp time arrays, in ecliptic inertial frame
+    combined_exptime_45 = np.zeros(num_el_bins_l1c * num_az_bins_l1c)
+    combined_exptime_90 = np.zeros(num_el_bins_l1c * num_az_bins_l1c)
+
+    # Create ecliptic inertial grid onto which we will project the individual DPS frames
+    # Build the azimuth and elevation grid in the heliocentric ecliptic frame (HAE)
+    (az_range, el_range, az_grid, el_grid) = l2_utils.build_az_el_grid(
+        spacing=spacing_deg,
+        input_degrees=True,
+        output_degrees=False,
+        centered_azimuth=False,  # (0, 2pi rad = 0, 360 deg)
+        centered_elevation=True,  # (-pi/2, pi/2 rad = -90, 90 deg)
+    )
+
+    for prod_num, l1c_prod in enumerate(l1c_products):
+        time = float(l1c_prod.epoch.values)
+        logger.info(
+            f"Projecting exposure time from product at time index {prod_num} "
+            f"with epoch {time} = {spice.et2utc(time, 'ISOC', 3)}"
+        )
+        # TODO: Reset the frame definition for the appropriate Pointing
+
+        (
+            az_grid_raveled,
+            el_grid_raveled,
+            flat_indices_helio,
+            flat_indices_dps,
+            hae_az_in_dps_az,
+            hae_el_in_dps_el,
+            hae_az_in_dps_az_indices,
+            hae_el_in_dps_el_indices,
+        ) = project_inertial_frame_to_dps(
+            event_time=time,
+            existing_grids=(az_grid, el_grid),
+            spacing_deg=spacing_deg,
+        )
+
+        # Add to the combined exposure time arrays
+        pointing_projected_exptime = l1c_prod.exposure_time.values.ravel(order="F")[
+            flat_indices_dps
+        ]
+        if is_ultra45(l1c_prod):
+            combined_exptime_45 += pointing_projected_exptime
+        else:
+            combined_exptime_90 += pointing_projected_exptime
+
+    combined_exptime_total = combined_exptime_45 + combined_exptime_90
+
+    return (combined_exptime_45, combined_exptime_90, combined_exptime_total)
+
+
+@typing.no_type_check
+@ensure_spice
```

Comment:
Same comment as above. Unless you are directly calling a spiceypy function, within this function, you do not need these decorators.